### PR TITLE
Add W&B backend for hyperparameter sweep

### PR DESCRIPTION
**Pull request by AyushExel**
# Add support for W&B hyperparameter sweep
This PR:
* allows using wandb for running hyperparameter search.
* The runs are visualized on W&B sweeps dashboard
* This supports runnning sweeps on parallel devices, all reporting to the same central dashboard.

### Usage
**To run new a hyperparameter search:**
```
trainer.hyperparameter_search(
    backend="wandb", 
    project="transformers_sweep", # name of the project
    n_trials=5,
    metric="eval/loss", # metric to be optimized, default 'eval/loss'. A warning is raised if the passed metric is not found
)
```
This outputs a sweep id. Eg. `my_project/sweep_id`

**To run sweeps on parallel devices:**
Just pass sweep id which you want to run parallel
```
trainer.hyperparameter_search(
    backend="wandb", 
    sweep_id = "my_project/sweep_id"
)
```
## Example sweep [[Dashboard](https://wandb.ai/cayush/teansformers_sweep/sweeps/xrmhkad5?workspace=user-cayush)] - runs from 2 devices were reported in this dashboard.
Screenshot from 2021-12-01 14-14-09

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### Regarding tests
the testing suite didn't include any tests as it is tested on wandb's side. Same can be done for the sweeps integrations

## Who can review?
Tagging Boris as he did the W&B integration and Sylvain as this PR introduces changes in integrations and trainer
@borisdayma @sgugger

### CODE REVIEW

File: src/transformers/integrations.py [hunk 1]
```diff
@@ -337,6 +341,74 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
     return best_run
 
 
+def run_hp_search_wandb(trainer, n_trials: int, direction: str, **kwargs) -> BestRun:
+    from .integrations import is_wandb_available
+
+    assert is_wandb_available(), "This function needs wandb installed: `pip " "install wandb`"
```
Comment (sgugger) [inline src/transformers/integrations.py:347]: Please use an if and raise an `ImportError` here, we don't accept new `assert` in the code base :-)

File: src/transformers/integrations.py [hunk 2]
```diff
@@ -337,6 +341,74 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
...
+            break
+    if not reporting_to_wandb:
+        trainer.add_callback(WandbCallback())
+    trainer.args.report_to = "wandb"
+    best_trial = {"run_id": None, "objective": None, "hyperparameters": None}
+    sweep_id = kwargs.pop("sweep_id", None)
+    project = kwargs.pop("project", None)
+    name = kwargs.pop("name", None)
+    entity = kwargs.pop("entity", None)
+    metric = kwargs.pop("metric", None) or "eval/loss"
```
Comment (sgugger) [inline src/transformers/integrations.py:364]: Why not put it directly as a default value?
```suggestion
    metric = kwargs.pop("metric", "eval/loss")
```

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -869,7 +871,6 @@ def num_examples(self, dataloader: DataLoader) -> int:
     def _hp_search_setup(self, trial: Union["optuna.Trial", Dict[str, Any]]):
         """HP search setup code"""
         self._trial = trial
-
```
Comment (sgugger) [inline src/transformers/trainer.py:872]: Let's leave the white lines pretty please :-)

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -2145,15 +2157,12 @@ def evaluate(
         )
 
         self.log(output.metrics)
-
```
Comment (sgugger) [inline src/transformers/trainer.py:2148]: Same here and for the two below.

File: src/transformers/trainer_utils.py [hunk 5]
```diff
@@ -211,16 +211,35 @@ def default_hp_space_sigopt(trial):
     ]
 
 
+def default_hp_space_wandb(trial) -> Dict[str, float]:
+    from .integrations import is_wandb_available
+
+    assert is_wandb_available(), "This function needs wandb installed: `pip " "install wandb`"
```
Comment (sgugger) [inline src/transformers/trainer_utils.py:217]: Same as above, please use an if and raise an ImportError.

**Review by sgugger — COMMENTED** [review]
Thanks a lot for your PR! I just a few styling comments and this should be good to merge.

File: src/transformers/integrations.py (hunk 1 above)
Comment (AyushExel) [reply src/transformers/integrations.py:347]: Makes sense. I'll make the change.

File: src/transformers/integrations.py (hunk 2 above)
Comment (AyushExel) [reply src/transformers/integrations.py:364]: lol yes! will do

File: src/transformers/trainer.py (hunk 3 above)
Comment (AyushExel) [reply src/transformers/trainer.py:872]: will do!

File: src/transformers/trainer.py (hunk 4 above)
Comment (AyushExel) [reply src/transformers/trainer.py]: sure

**Comment by LysandreJik** [comment]
Thanks for your PR! I will take a look at your PR and help out on tests after the holiday break - first week of January! Thanks for your patience.

**Comment by AyushExel** [comment]
@LysandreJik Hey! any updates on this?

**Comment by LysandreJik** [comment]
Thank you for your PR, @AyushExel! All that remains is testing.

You can see an example of how it was done for `ray` and `sigopt` here: https://github.com/huggingface/transformers/blob/f21bc4215aa979a5f11a4988600bc84ad96bef5f/tests/test_trainer.py#L1582-L1692

Ideally, we would move all of that to an `integrations` folder in the [tests](https://github.com/huggingface/transformers/tree/master/tests) folder. Then we could add a `test_wandb_integration.py` which would contain a similar integration test to those above.

The idea is that when the test fail, we ping the author(s) of the PRs, this way we share the burden of the maintenance. 

Does that work for you?

**Comment by AyushExel** [comment]
@LysandreJik sounds good!
I've added the tests to the same file `test_trainer.py` for now. Do you want me to move it out to integrations before merging or is that a future project?

**Comment by sgugger** [comment]
Thanks for adding the test! Can it run out of the box if wandb is installed? If I recall correctly there is some login to do no?

**Comment by AyushExel** [comment]
@sgugger I've kept anonymous mode on by default in the test so it should run. Please, Let me know if that's not the case and I'll think of a workaround

**Comment by AyushExel** [comment]
@LysandreJik I guess there are 2 options here.
* Add wandb api key as a github secret and run `wandb login $secret_key` in the CI test workflow setup
* Or Setup a dummy account meant only for test and login directly `wandb login key` without having to worry about exposing the api key

**Comment by AyushExel** [comment]
Hey @LysandreJik were you able to run the tests?

**Comment by LysandreJik** [comment]
Hey @AyushExel, I think we can go with the first solution:
- Add wandb api key as a github secret and run wandb login $secret_key in the CI test workflow setup

Would you be able to share an API key we could register in secrets to get this working?

**Comment by AyushExel** [comment]
@LysandreJik each wandb account has an api key which is accessible at [wandb.ai/authorize](https://wandb.ai/authorize)

**Comment by morganmcg1** [comment]
@LysandreJik it probably makes more sense for you to create a dummy wandb account with a hugging face email that you control and use this accounts API key :)

**Comment by LysandreJik** [comment]
Ok will give it a look!

**Comment by LysandreJik** [comment]
I've added a key for a dummy account in the `WANDB_API_KEY` secret of this repo. Is it possible for you to leverage it in Github Actions and to pass it to this test in your PR?

**Comment by AyushExel** [comment]
@LysandreJik Yes..In github actions you can run `wandb login YOUR_SECRET_ WANDB_API_KEY` ,it will log you in on that system and the tests will execute fine

**Comment by LysandreJik** [comment]
Right, I was asking if you could please add it to the GitHub action workflows in your PR. I've added the secret, you should be able to handle it from your PR now. 

The jobs that run with integrations are the following:
- https://github.com/huggingface/transformers/blob/master/.github/workflows/self-scheduled.yml#L21
- https://github.com/huggingface/transformers/blob/master/.github/workflows/self-scheduled.yml#L253

If this is not possible, then I will open a PR on your PR.

Thank you.

**Comment by AyushExel** [comment]
Okay got it.. I'm trying to set it up

**Comment by AyushExel** [comment]
@LysandreJik I've added the wandb login command in both the tests you mentioned. Can you please take a look?

**Review by LysandreJik — APPROVED** [review]
Great, thanks for adding that! Let's merge it like that and I'll monitor in the coming days :)

Thanks!

**Pull request merged**